packages/core/src/stateUtils.ts

```diff
@@ -259,15 +259,49 @@ export const isStateId = (str: string) => str[0] === STATE_IDENTIFIER;
 
 export function getCandidates<TEvent extends EventObject>(
   stateNode: StateNode<any, TEvent>,
-  eventName: TEvent['type'] | NullEvent['type'] | '*'
+  eventName: TEvent['type'] | NullEvent['type'],
+  /**
+   * If `true`, will use SCXML event token matching semantics
+   */
+  matchTokens: boolean = false
 ): Array<TransitionDefinition<any, TEvent>> {
   const transient = eventName === NULL_EVENT;
   const candidates = stateNode.transitions.filter((transition) => {
-    const sameEventType = transition.eventType === eventName;
-    // null events should only match against eventless transitions
-    return transient
-      ? sameEventType
-      : sameEventType || transition.eventType === WILDCARD;
+    // First, check the trivial case: event names are exactly equal
+    if (transition.eventType === eventName) {
+      return true;
+    }
+
+    // Then, check if transition is a wildcard transition,
+    // which matches all non-transient events
+    if (transition.eventType === WILDCARD) {
+      return !transient;
+    }
```

Comment:
This works OK now but I had to spend a little bit of time to see if everything is OK here - especially given the fact that the `eventType: '.*'` gets processed by the rest of this function. I think it's an incorrect descriptor and it's correctly not matched here but I think that this would be slightly more obvious to analyze:
```suggestion
    // transient transitions can't match non-transient events
    if (transient) {
        return false;
    }
    // Then, check if transition is a wildcard transition,
    // which matches any non-transient events
    if (transition.eventType === WILDCARD) {
      return true;
    }
```